### ISSUE TITLE
Unsupported Load Command LC_DYLD_CHAINED_FIXUPS

### ISSUE THREAD

**Issue by fmagin** [NationalSecurityAgency/ghidra#3572]
**Describe the bug**
Ghidra does not handle the new MachO Load Commands that are used in iOS Apps targeting iOS15 and above.
Specifically:
`0x80000034` which is `LC_DYLD_CHAINED_FIXUPS` which is essential for correctly loading binaries, and correctly rebasing internal pointers. 
`0x80000033` which is `LC_DYLD_EXPORTS_TRIE`. This is related to the exports of the binary, and doesn't cause much issues I am aware of, 

**To Reproduce**
Load the attached binary

**Expected behavior**

The binary being loaded correctly, pointers being fixed and symbols being added

**Attachments**

Simple example binary that uses the load commands:
[iOS15_example.zip](https://github.com/NationalSecurityAgency/ghidra/files/7448875/iOS15_example.zip)

**Environment (please complete the following information):**
 - OS: Linux
 - Java Version: 11.0.12
 - Ghidra Version: 10.0.4
 - Ghidra Origin: Github Release Asset

**Additional context**

Not many tools handle the new load commands correctly yet. IDA Pro technically has support, but this involves asking the support and getting a patched macho loader library.
[datatheorem/strongarm](https://github.com/datatheorem/strongarm) handles it and includes an [explanation](https://github.com/datatheorem/strongarm/blob/release/chained_fixup_pointers.md) what the new load commands do.

For a lightweight tool to just look at the basic results https://github.com/horsicq/XMachOViewer seems to work well enough.

**Comment by fmagin** [comment]
I think the relevant logic is implemented by ghidra.app.util.opinion.DyldCacheProgramBuilder#fixPageChains and ghidra.app.util.opinion.DyldCacheProgramBuilder#processPointerChain already. This just needs to be applied to the pages specified in the new `LC_DYLD_CHAINED_FIXUPS` load command.

**Comment by ryanmkurtz** [comment]
We have been working on supporting this for 10.1.

**Comment by ryanmkurtz** [comment]
Can you give these changes a try?  We'd like to work out any kinks before the 10.1 release.

**Comment by fmagin** [comment]
The fixups themselves seem to work, or at least happen, though the subsequent ObjC analysis still don't recover the classes. The output of the loader is:
```
Skipping segment: __PAGEZERO (nslogtest2)
Skipping section: __DATA.__objc_data (nslogtest2)
Unsupported Load Command Type = 0x2c
Fixed 2 PTR_64_OFFSET chained pointers.
Fixed 1 PTR_64_OFFSET chained pointers.
Fixed up 3 chained pointers.
```
`Skipping section: __DATA.__objc_data ` sounds fairly suspicious.

At address `0x1000082d8` at the start of the `__objc_classlist` section is the value `0000c0e8` which seems like a pointer that wasn't rebased, which might be while the ObjC class analysis doesn't recover anything.

![image](https://user-images.githubusercontent.com/8415354/140306622-0db83181-0448-4b6c-b0ed-9f619098a6b5.png)

One thing that definitely changed with iOS 15 is that the method references inside the ObjC data are now relative pointers, but this should only crop up as an error later, the rest of the data structures are the same IIRC. So there is most likely something going wrong there in the loader itself.



There are also some other issues that I haven't opened yet, like `LC_FUNCTION_STARTS` not being used for function creation (which makes Ghidra miss various functions), so this is definitely something else I'd really like to see in 10.1.

**Comment by ryanmkurtz** [comment]
Ok thanks.  Yes if you could make new issues for all the problems you discover it would be appreciated.

**Comment by fmagin** [comment]
Just checked the binary with [datatheorem/strongarm](https://github.com/datatheorem/strongarm) and this finds:
`0x82d8: DyldChainedPtr64Rebase(raw: 0xc0e8) target=Phys[0xc0e8]`. I.e. the pointer in the screenshot above should have been fixed up to `0x10000c0e8`. So I would consider this issue not closed, because https://github.com/NationalSecurityAgency/ghidra/commit/e8550982f30f655d9d0fbc2b96ebfd4320b3bf40 doesn't seem to properly fix all pointers.

**Comment by fmagin** [comment]
Overall Ghidra finds 3 chained pointers, while strongarm finds 22:
```
{0x100008100: 0x100007a40,
 0x1000081e0: 0x1000072b0,
 0x1000081e8: 0x1000072dc,
 0x1000081f0: 0x1000072e4,
 0x1000081f8: 0x100007310,
 0x100008200: 0x100007350,
 0x100008208: 0x10000735c,
 0x100008210: 0x10000738c,
 0x100008218: 0x1000073e8,
 0x100008238: 0x1000081e0,
 0x100008248: 0x100007a40,
 0x1000082a0: 0x100007afc,
 0x1000082d8: 0x10000c0e8,
 0x10000c018: 0x100007b90,
 0x10000c060: 0x100007b90,
 0x10000c068: 0x100007998,
 0x10000c090: 0x100007c2a,
 0x10000c0d0: 0x10000c000,
 0x10000c0d8: 0x100007210,
 0x10000c0e8: 0x10000c0b0,
 0x10000c108: 0x10000c04a,
 0x10000c128: 0x1000079f0}
```

**Comment by mumbel** [comment]
@emteere  would doing stuff in MachoProgramBuilder.markupHeaders() (or just in MachoProgramBuilder in general) be needed/helpful, or is that just small struct markup and not much value add.  Didnt know if that adds alot just by adding another `else if` for instance of the chained command... and/or linkedit command

**Comment by ryanmkurtz** [comment]
For the most part all the Mach-O structures are laid down in their corresponding class's `toDataType()` methods.  The stuff being done in `markupHeaders()` is typically just for the dynamic stuff that `toDataType()` couldn't handle, like variable length strings.

**Comment by mumbel** [comment]
Ah, this seemed like the place that `dyld_chained_fixups_header` would get addressed, (I didn't see this struct applied at least, but that might just be a value add question), I was just tracking down other places that toDataType and createData got called in that code, I guess I missed the static stuff

**Comment by fmagin** [comment]
As far as I see symbol binding is also not handled:
![image](https://user-images.githubusercontent.com/8415354/140391983-ce9988fa-c458-4700-b813-5f0fced00d22.png)

`10000c090` should have a rebased pointer (see dictionary posted earlier), but `10000c098` and `10000c0a0` should also contain pointers to `_OBJC_CLASS_$_OS_os_log` and `_OBJC_CLASS_$_NSString` respectively according to  [datatheorem/strongarm](https://github.com/datatheorem/strongarm).

**Comment by emteere** [comment]
Thanks for the sample.  I was able to find 2 bugs for the PTR_64_OFFSET type chains (and 32 bit chains as well).

The first, when extracting the next pointer value, the code was shifting by 52, when it should have been shifting by 51.  So the next value stopped the chain at the first entry in the in chain.

For the second bug, the stride offset for the first entry in a chain, which is byte based, was always using 8, instead of the correct stride.

With the two corrections, there are now 105 chain fixup entries fixed, including the ones you mention above.

I'll get the fix in tomorrow

**Comment by ryanmkurtz** [comment]
Fixes are in bda7e0df453fca4c8df51e898404ed6e5314ec61

**Comment by ryanmkurtz** [comment]
> Ah, this seemed like the place that `dyld_chained_fixups_header` would get addressed, (I didn't see this struct applied at least, but that might just be a value add question), I was just tracking down other places that toDataType and createData got called in that code, I guess I missed the static stuff

@mumbel i think i misunderstood your original comment about markup.  I just implemented LC_FUNCTION_STARTS and yea, it doesn't seem like we currently have a good place to markup what these load commands point to.  Our design accounts for marking them up when you import with the BinaryLoader, but not the MachoLoader.
